lib/tesla/middleware/path_params.ex

```diff
@@ -34,9 +54,50 @@ defmodule Tesla.Middleware.PathParams do
 
   defp build_url(url, nil), do: url
 
+  defp build_url(url, %_{} = params), do: build_url(url, Map.from_struct(params))
+
   defp build_url(url, params) do
-    Regex.replace(@rx, url, fn match, key ->
-      to_string(params[String.to_existing_atom(key)] || match)
-    end)
+    params
+    |> Enum.map(fn {name, value} -> {to_string(name), value} end)
+    |> Enum.sort_by(fn {name, _} -> String.length(name) end, :desc)
+    |> Enum.reduce(url, &replace_parameter/2)
   end
+
+  # Do not replace parameters with nil values.
+  defp replace_parameter({_name, nil}, url), do: url
+  defp replace_parameter({name, value}, url), do: replace_parameter(to_string(name), value, url)
+
+  defp replace_parameter(<<h, t::binary>>, value, url)
+       when h in ?a..?z or h in ?A..?Z do
+    case parse_name(t, <<h>>) do
+      :error ->
+        url
+
+      name ->
+        encoded_value =
+          value
+          |> to_string()
+          |> URI.encode_www_form()
+
+        url
+        |> String.replace(":#{name}", encoded_value)
+        |> String.replace("{#{name}}", encoded_value)
```

Comment:
I wonder if we should pick one or another based on a configuration. My stand is to wait until somebody complain being "slow", but I would like to hear your thoughts about it

Reply:
If we wanted to separate it — and I’m not sure with the most recent commit that it is necessary — I would have something like `OpenApiPathParams` and leave phoenix style parameters to `PathParam`. I would probably keep a mostly-common implementation because this *also* fixes #566.

Reply:
I would lean towards adding `OpenApiPathParams` module honestly! I am curious to see how simple could be based on its constraints

Reply:
Hm. The main difference would be instead of using one regex (`~r/PHOENIX_STYLE|OPENAPI_STYLE/`) there would be separate regexes. The `Regex.replace/4` call would use the appropriate regex for its type and instead of having to know `match, "", openapi` and `match, phoenix, _openapi`, we would just be doing `match,  name`.

So not *that* much simpler than what we have right now.